tests/integration/container_engine.py

```diff
@@ -92,11 +94,139 @@ def run_cmd_on_image(
+        image_cmd, image_entrypoint = self._get_image_config(image)
+
+        # fallback to the image's default cmd if a custom one is not provided
+        cmd = cmd or image_cmd
+
+        # if an entrypoint flag is provided, use it
+        if entrypoint:
+            return ["run", *filtered_flags, container_name, "--", entrypoint, *cmd]
+
+        # if the image has an entrypoint, prepend it to the command
+        if image_entrypoint:
+            return ["run", *filtered_flags, container_name, "--", *image_entrypoint, *cmd]
+
+        return ["run", *filtered_flags, container_name, "--", *cmd]
+
+    def _filter_entrypoint_flags(
+        self, flags: Optional[list[str]]
+    ) -> tuple[list[str], Optional[str]]:
+        """Remove and return any entrypoint flags.
+
+        Since buildah does not support the --entrypoint flag, we need to extract it so that
+        the run command can be properly formatted later on.
+
+        :param flags: The flags to treat.
+        :return: The treated flags and the entrypoint value.
+        """
+        if flags is None:
+            return [], None
+
+        # find the index of the entrypoint flag, if it exists
+        index = next((i for i, flag in enumerate(flags) if flag.startswith("--entrypoint")), None)
+
+        if index is None:
+            return flags, None
+
+        entrypoint_flag = flags[index]
+
+        # entrypoint flag is in the format --entrypoint=<entrypoint>
+        if "=" in entrypoint_flag:
+            entrypoint_value = entrypoint_flag.split("=")[1]
+            return flags[:index] + flags[index + 1 :], entrypoint_value
+        # entrypoint flag is in the format --entrypoint <entrypoint>
+        elif index + 1 < len(flags):
+            entrypoint_value = flags[index + 1]
+            return flags[:index] + flags[index + 2 :], entrypoint_value
+        # entrypoint flag exists but no value is provided
+        else:
+            raise RuntimeError(
+                f"Unable to parse the run command flags: {flags}."
+                "Check if the --entrypoint flag defines a value."
+            )
+
+    def _get_image_config(self, image: str) -> tuple[list[str], list[str]]:
+        """Get the configured entrypoint and cmd of an image."""
+        output, exit_code = self._run_cmd(["buildah", "inspect", image])
+
+        if exit_code != 0:
+            raise RuntimeError(f"Failed to inspect image {image}.")
+
+        parsed_output = json.loads(output)
+        docker_config = parsed_output.get("Docker", {}).get("config", {})
+        cmd = docker_config.get("Cmd", [])
+        entrypoint = docker_config.get("Entrypoint", [])
+
+        return cmd, entrypoint
+
+    def run_cmd_on_image(
+        self,
+        image: str,
+        cmd: list[str],
+        flags: Optional[list[str]] = None,
+    ) -> tuple[str, int]:
+        """Run command using buildah."""
+        container_name = self._configure_buildah_container(image)
```

Comment:
I wonder if it makes sense to lift `finally` into `_configure_buildah_container` and turn it into a context manager? It could yield a container name and then dispose of the container. Or would that mess with any other intended use?

Reply:
Yeah, good suggestion.